### ISSUE TITLE
Fix Shift+N behavior in v1.3-beta

### ISSUE THREAD

**Issue by dacap** [aseprite/aseprite#2957]
Reported here: https://community.aseprite.org/t/add-new-layer-broken/10760

<kbd>Shift+N</kbd> should work exactly as in Aseprite v1.2.28 (`main` branch).

**Comment by iamOgunyinka** [comment]
Here's what my research reveals.

When the focus is in the app editor and the `Shift` is pressed. We enter a [StandbyState](https://github.com/aseprite/aseprite/blob/f318a42f199302b1f54bf7a1438e95d7486a4584/src/app/ui/editor/standby_state.cpp#L514) and this creates a [drawing state](https://github.com/aseprite/aseprite/blob/f318a42f199302b1f54bf7a1438e95d7486a4584/src/app/ui/editor/standby_state.cpp#L670). The drawing state after series of calls creates a transaction and [increments the Undo counter](https://github.com/aseprite/aseprite/blob/f318a42f199302b1f54bf7a1438e95d7486a4584/src/app/cmd_transaction.cpp#L104). Although there are visible drawings in the editor, the changes are not committed until we key-up/drop the pressed key.

Combining the `Shift` key with key `N` creates a totally new command ([the new layer](https://github.com/aseprite/aseprite/blob/f318a42f199302b1f54bf7a1438e95d7486a4584/src/app/commands/cmd_new_layer.cpp#L96)) and executing it. Before new commands are run, we needed to signal all interested "candidates" [for a before](https://github.com/aseprite/aseprite/blob/f318a42f199302b1f54bf7a1438e95d7486a4584/src/app/context.cpp#L182) and [after](https://github.com/aseprite/aseprite/blob/f318a42f199302b1f54bf7a1438e95d7486a4584/src/app/context.cpp#L195) command execute state change. In this case, two of such "interested candidates" in the `BeforeCommandExecution` signal is the [Timeline](https://github.com/aseprite/aseprite/blob/f318a42f199302b1f54bf7a1438e95d7486a4584/src/app/ui/timeline/timeline.cpp#L264-L267) and the [DrawingState](https://github.com/aseprite/aseprite/blob/f318a42f199302b1f54bf7a1438e95d7486a4584/src/app/ui/editor/drawing_state.cpp#L56). These two (along with possibly some others) are signaled in the order by which they connected to that signal. The `Timeline` is signaled first and it [caches the value of the Undo counter](https://github.com/aseprite/aseprite/blob/f318a42f199302b1f54bf7a1438e95d7486a4584/src/app/ui/timeline/timeline.cpp#L1785). Then the DrawingState [too gets its turn](https://github.com/aseprite/aseprite/blob/f318a42f199302b1f54bf7a1438e95d7486a4584/src/app/ui/editor/drawing_state.cpp#L373).

The drawing state that was created above, needs to either commit its previous changes or roll it back. In this case, we never committed the straight line drawing that the `Shift` key creates, so we had to [rollback to the last "stable" state](https://github.com/aseprite/aseprite/blob/f318a42f199302b1f54bf7a1438e95d7486a4584/src/app/ui/editor/drawing_state.cpp#L395), that means decrementing the [Undo counter that was incremented earlier](https://github.com/aseprite/aseprite/blob/f318a42f199302b1f54bf7a1438e95d7486a4584/src/app/cmd_transaction.cpp#L112).

Then `CmdTransaction::onExecute()` then gets executed. This [increments the counter](https://github.com/aseprite/aseprite/blob/f318a42f199302b1f54bf7a1438e95d7486a4584/src/app/cmd_transaction.cpp#L104) again. 

Then the `AfterCommandExecution` signal is emitted. The `Timeline` [checks if the current value of the cached counter has changed](https://github.com/aseprite/aseprite/blob/f318a42f199302b1f54bf7a1438e95d7486a4584/src/app/ui/timeline/timeline.cpp#L1795-L1796). If it has, it refreshes the timeline. Otherwise, it's a no-op. In this case, the value of the counter remains the same and thus, a stale timeline that didn't get updated with the _possibly_ restacked layers.


- tl;dr
In the case where a new layer is added when the focus is in the editor, this bug is going to appear due to the drawing state decrementing the shared state counter after it has been cached by the timeline. 

- Possible fix?
We can move the [regenerateRows](https://github.com/aseprite/aseprite/blob/f318a42f199302b1f54bf7a1438e95d7486a4584/src/app/ui/timeline/timeline.cpp#L1797) outside of the check. This is *a* fix but not *the* fix.

**Comment by dacap** [comment]
Thanks for the analysis @iamOgunyinka, seeing the code it looks like the [`m_cmds->execute()` call from Transaction() ctor](https://github.com/aseprite/aseprite/blob/364f62ee4a74d85fecdda42ab9a601a54945d629/src/app/transaction.cpp#L55) was a hack that generated this issue indirectly.

I'll need some time to think about it. The bad thing is that the "saved counter" is incremented but the `Transaction` is still being executed (so a rollback would restore the "saved counter" as you saw).

One possible simpler solution (?) could be that the `Timeline` save/cache the version of the sprite (the `doc::Sprite::version()`) instead of the `DocUndo::m_savedCounter` (as when the layers structure is modified, the sprite version is modified).

**Comment by iamOgunyinka** [comment]
We discussed this some weeks ago, any solution yet?

**Comment by dacap** [comment]
Hi Josh! Maybe you can give a try to comparing sprite ID and sprite version instead of the saved counter (the timeline will need a couple of fields `m_spriteId` and `m_spriteVersion`). I'm not sure if this will solve the problem or if it will have the same problem as the undo counter, but you can give a try.
